api/types/entrySchema.go

```diff
@@ -0,0 +1,181 @@
+package apitypes
+
+import (
+	"encoding/json"
+	"fmt"
+	"sort"
+
+	"github.com/emirpasic/gods/maps/linkedhashmap"
+	"github.com/getlantern/deepcopy"
+	"github.com/puppetlabs/wash/plugin"
+)
+
+// EntrySchema describes an entry's schema, which
+// is what's returned by the /fs/schema endpoint.
+type EntrySchema struct {
+	plugin.EntrySchema
+	children map[string]*EntrySchema
+	// graph is an ordered map of <TypeID> => <EntrySchema>. We store it to make
+	// MarshalJSON's implementation easier.
+	//
+	// NOTE: The reason we don't synchronize children with graph is b/c entry
+	// schemas are immutable. Clients that want to mess with the state (e.g. like
+	// `wash find`) can do so at their own peril.
+	graph *linkedhashmap.Map
+}
+
+// MarshalJSON marshals the entry's schema to JSON. It takes
+// a value receiver so that the entry schema's still marshalled
+// when it's referenced as an interface{} object. See
+// https://stackoverflow.com/a/21394657 for more details.
+func (s EntrySchema) MarshalJSON() ([]byte, error) {
+	if s.graph != nil {
+		return s.graph.ToJSON()
+	}
+	return json.Marshal(s.EntrySchema)
+}
+
+// UnmarshalJSON unmarshals the entry's schema from JSON.
+func (s *EntrySchema) UnmarshalJSON(data []byte) error {
+	rawGraph := linkedhashmap.New()
+	if err := rawGraph.FromJSON(data); err != nil {
```

Comment:
Won't this error if `s.graph` was nil when you marshalled it?

Reply:
Not sure what you mean. If you're talking about data being the "null" string, then that's taken care of by making `EntrySchema` a pointer. This UnmarshalJSON's only called when data isn't "null".

Reply:
MarshalJSON has if s.graph != nil marshal the graph, else marshal s.EntrySchema. But UnmarshalJSON seems to assume that it marshaled a linkedhashmap no matter what. Is that because plugin.EntrySchema uses the same marshaling? If so why do we have two different types to store it in?

Reply:
Yeah, plugin.EntrySchema [creates a linkedhashmap and marshals that](https://github.com/ekinanp/wash/blob/schema-cleanup/plugin/entrySchema.go#L99-L105). 

Reason for the different types is b/c plugin.EntrySchema's meant to be used by plugin authors for constructing entry schemas. You cannot do more complicated schema operations like graph-traversal or schema-pruning (those would require more documentation in godocs for people to sift through). All the more complicated schema-stuff is handled in the apitypes.EntrySchema class (I didn't add any graph-traversal functions yet because I wasn't sure how to abstract out some of the wash find/stree traversal since those traversals required knowledge of the children's results and passing in multiple callbacks to a generic function seemed complicated).

Reply:
I think I might remove `SetLabel` per https://github.com/puppetlabs/wash/pull/377/commits/670d53ae10f56f541c383b9977f02d902d2ee288 since that's basically saying SetLabel doesn't properly work without some more code.